src/librbd/mirror_snapshot/CreatePrimaryRequest.cc

```diff
@@ -0,0 +1,335 @@
+
+  librados::AioCompletion *comp = create_rados_callback<
+    CreatePrimaryRequest<I>, &CreatePrimaryRequest<I>::handle_get_mirror_peers>(this);
+  m_out_bl.clear();
+  int r = m_image_ctx->md_ctx.aio_operate(RBD_MIRRORING, comp, &op, &m_out_bl);
+  ceph_assert(r == 0);
+  comp->release();
+}
+
+template <typename I>
+void CreatePrimaryRequest<I>::handle_get_mirror_peers(int r) {
+  CephContext *cct = m_image_ctx->cct;
+  ldout(cct, 20) << "r=" << r << dendl;
+
+  std::vector<cls::rbd::MirrorPeer> peers;
+  if (r == 0) {
+    auto iter = m_out_bl.cbegin();
+    r = cls_client::mirror_peer_list_finish(&iter, &peers);
+  }
+
+  if (r < 0) {
+    lderr(cct) << "failed to retrieve mirror peers: " << cpp_strerror(r)
+               << dendl;
+    finish(r);
+    return;
+  }
+
+  for (auto &peer : peers) {
+    if (peer.mirror_peer_direction == cls::rbd::MIRROR_PEER_DIRECTION_RX) {
+      continue;
+    }
+    m_mirror_peers.insert(peer.uuid);
+  }
+
+  if (m_mirror_peers.empty()) {
+    lderr(cct) << "no mirror tx peers configured for the pool" << dendl;
+    finish(-EINVAL);
+    return;
+  }
+
+  create_snapshot();
+}
+
+template <typename I>
+void CreatePrimaryRequest<I>::create_snapshot() {
+  CephContext *cct = m_image_ctx->cct;
+  ldout(cct, 20) << dendl;
+
+  cls::rbd::MirrorPrimarySnapshotNamespace ns{m_demoted, m_mirror_peers};
+  auto ctx = create_context_callback<
+    CreatePrimaryRequest<I>,
+    &CreatePrimaryRequest<I>::handle_create_snapshot>(this);
+  m_image_ctx->operations->snap_create(ns, m_snap_name, ctx);
+}
+
+template <typename I>
+void CreatePrimaryRequest<I>::handle_create_snapshot(int r) {
+  CephContext *cct = m_image_ctx->cct;
+  ldout(cct, 20) << "r=" << r << dendl;
+
+  if (r < 0) {
+    lderr(cct) << "failed to create mirror snapshot: " << cpp_strerror(r)
+               << dendl;
+    finish(r);
+    return;
+  }
+
+  if (m_snap_id != nullptr) {
+    std::shared_lock image_locker{m_image_ctx->image_lock};
+    cls::rbd::MirrorPrimarySnapshotNamespace ns{m_demoted, m_mirror_peers};
+    *m_snap_id = m_image_ctx->get_snap_id(ns, m_snap_name);
+  }
+
+  unlink_peer();
+}
+
+template <typename I>
+void CreatePrimaryRequest<I>::unlink_peer() {
+  uint64_t max_snapshots = m_image_ctx->config.template get_val<uint64_t>(
+    "rbd_mirroring_max_mirroring_snapshots");
+  ceph_assert(max_snapshots >= 3);
+
+  std::string peer_uuid;
+  uint64_t snap_id = CEPH_NOSNAP;
+
+  for (auto &peer : m_mirror_peers) {
+    std::shared_lock image_locker{m_image_ctx->image_lock};
+    size_t count = 0;
+    uint64_t prev_snap_id = 0;
+    for (auto &snap_it : m_image_ctx->snap_info) {
+      if (boost::get<cls::rbd::MirrorNonPrimarySnapshotNamespace>(
+            &snap_it.second.snap_namespace)) {
+        break;
+      }
+      auto info = boost::get<cls::rbd::MirrorPrimarySnapshotNamespace>(
+        &snap_it.second.snap_namespace);
+      if (info == nullptr) {
+        continue;
+      }
+      if (info->demoted) {
```

Comment:
@dillaman  At the moment I am no sure this sequence is possible at all. Because before creating NP1 the rbd-mirror seems can safely delete P1, and when P2-NP1 diff is copied it can delete P2, and it can't create P3 if NP2 is not fully copied (in this case I suppose the force promotion would mean rollback to (fully copied) NP1 first).

Anyway, it seems instead of breaking here it would be better just to reset the counter, so we are always counting only snapshots after the latest promotion, expecting that previous will be eventually deleted by the rbd-mirror. Do you like this? And I think it can always be improved later when we have something working.